pkg/providers/vsphere/vsphere.go

```diff
@@ -443,6 +450,80 @@ func (p *vsphereProvider) validateMachineConfigsNameUniqueness(ctx context.Conte
 	return nil
 }
 
+func (p *vsphereProvider) getCurrentMachineConfigDatastoreUsage(ctx context.Context, machineConfig *v1alpha1.VSphereMachineConfig, vsphereClusterSpec *Spec, cluster *types.Cluster) (diskGiB float64, err error) {
+	if cluster == nil {
+		return 0, nil
+	}
+	em, err := p.providerKubectlClient.SearchVsphereMachineConfig(ctx, machineConfig.Name, cluster.KubeconfigFile, vsphereClusterSpec.Cluster.GetNamespace())
+	if err != nil {
+		return 0, err
+	}
+	if len(em) > 0 {
+		return float64(em[0].Spec.DiskGiB), nil
+	}
+	return 0, nil
+}
+
+func (p *vsphereProvider) calculateDatastoreUsage(ctx context.Context, machineConfig *v1alpha1.VSphereMachineConfig, vsphereClusterSpec *Spec, cluster *types.Cluster, usage map[string]*datastoreUsage) error {
+	availableSpace, needGiB, err := p.getMachineConfigDatastoreRequirements(ctx, machineConfig, vsphereClusterSpec.Cluster.Spec.ControlPlaneConfiguration.Count)
+	if err != nil {
+		return err
+	}
+	currentUsage, err := p.getCurrentMachineConfigDatastoreUsage(ctx, machineConfig, vsphereClusterSpec, cluster)
+	if err != nil {
+		return err
+	}
+	availableSpace += currentUsage
+	if _, ok := usage[machineConfig.Spec.Datastore]; ok {
+		usage[machineConfig.Spec.Datastore].needGiBSpace += needGiB
+		usage[machineConfig.Spec.Datastore].availableSpace += availableSpace
+	} else {
+		usage[machineConfig.Spec.Datastore] = &datastoreUsage{
+			availableSpace: availableSpace,
+			needGiBSpace:   needGiB,
+		}
+	}
+	return nil
+}
+
+func (p *vsphereProvider) getMachineConfigDatastoreRequirements(ctx context.Context, machineConfig *v1alpha1.VSphereMachineConfig, count int) (available float64, need int, err error) {
+	availableSpace, err := p.providerGovcClient.GetWorkloadAvailableSpace(ctx, machineConfig.Spec.Datastore) // TODO: remove dependency on machineConfig
+	if err != nil {
+		return 0, 0, fmt.Errorf("getting datastore details: %v", err)
+	}
+	needGiB := machineConfig.Spec.DiskGiB * count
+	return availableSpace, needGiB, nil
+}
+
+func (p *vsphereProvider) validateDatastoreUsage(ctx context.Context, vsphereClusterSpec *Spec, cluster *types.Cluster) error {
```

Comment:
We need the `kubectl` client to search for the current machine config objects on the cluster during upgrades. In the vsphere validator, we only make external calls using GoVC. Any validations that require the kubectl client are defined at the provider level & called from there.

Reply:
then should we add the kubectl client to the validator?

Reply:
I figured we were explicitly not defining it there since we separate validating machine config name uniqueness due to the dependency of kubectl client. https://github.com/aws/eks-anywhere/blob/main/pkg/providers/vsphere/vsphere.go#L413

If there are no foreseen issues with adding the kubectl client here, then i can make those changes

Reply:
ah you are right, i think the reason not injecting kubectl to validator is bc controller code uses the validation methods and can't use the same `kubectl` client without refactoring